In order to successfully complete this assignment we recommend that you participate both individually and in groups during class. **Turn in your assignment using D2L no later than 11:59pm on the day of class.** Grading is based on correctness and completion.

# 14 In-Class Assignment:  Page Rank (a.k.a. How Google Works)

    

### Agenda for today's class (80 minutes)

1. [(40 minutes) Pre-class assignment review](#Review)
1. [(40 minutes) Page Rank](#Page-Rank)

---
<a name="Review"></a>

## 1. Questions Regarding HW and Quiz 3 Material

This is your time to ask any questions regarding HW mistakes and quiz 3 material.
You can find the **Learning Goals for Homework 3 and Quiz 3 Topic** listed 
<a rel="Golas" href="https://docs.google.com/document/d/1x97QZ7nQTQHAhv_zJ6eabS1tnZRH9B7c0MYlvs0h_jY/edit?usp=sharing ">HERE</a>.


---
<a name="Page-Rank"></a>
## 2. Page Rank

Google's famous "Page Rank" algorithm is just a calculation of the likelyhood that a person will reach a particulare web page.  

> To explain the algorithm, we model the web as an oriented graph. The vertices are the pages, and the oriented edges are the links between pages. As we just explained, each page corresponds to a different url. Hence, a website may contain many pages. The model makes no difference between the individual pages of a website and its front page. But, most likely, the algorithm will rank better the front page of an important website.

From: http://blog.kleinproject.org/?p=280

Using the vocabulary from this course, A simplified version of Google's Page Rank is the steady state probabilities of the Markov State Model where each node in the network is a page and each transition probability is calculated as an equal probablility from the current page to its linked pages.  

For this problem, consider the following Markov model of a generic webpage. 

<img src="https://lh4.googleusercontent.com/LFe_yWxZA_F_Ze3PSMqwQds9zUOgjnzGP6pxPZJYAGjRTj_BKNqvcDxGHhbZg1Y3iSVOFPb9MW_a8JvwrdhfU3Av7zC13JO_faPs0WC7AziKWMwrwKkqdl6RJmxq=w477"></a>

&#9989; <font color=red>**Question 1.**</font> Find the state transition matrix for the above system. Let's consider Home = state 1, About = state 2, Resume = state 2, Blog = state 4 and let $p_{ij}$ be the probability transition from state $j$ to state $i$. 

In [2]:
%matplotlib inline 
import matplotlib.pylab as plt
import numpy as np
import sympy as sym
sym.init_printing()

A = np.matrix([[0.25,0.2,0.25,0.3],[0.5,0.5,0,0], [0.2,0,0.8,0], [0.1,0,0,0.9]]).T
sym.Matrix(A)

⎡0.25  0.5  0.2  0.1⎤
⎢                   ⎥
⎢0.2   0.5  0.0  0.0⎥
⎢                   ⎥
⎢0.25  0.0  0.8  0.0⎥
⎢                   ⎥
⎣0.3   0.0  0.0  0.9⎦

&#9989; <font color=red>**Question 2:**</font> Using ```numpy``` calculate the eigenvalues and eigenvectors of the model.

In [3]:
#put your answer here
#Put your answer here
eigval, eigvecs = np.linalg.eig(A)
print((eigval[0]), ",", (eigval[1]), ",", (eigval[2]), ",", eigval[3])
v1 = eigvecs[:,0]
v2 = eigvecs[:,1]
v3 = eigvecs[:,2]
v4 = eigvecs[:,3]
print(v1)
print()
print(v2)
print()
print(v3)
print()
print(v4)

-0.030805589196661765 , 0.6282457190832064 , 0.9999999999999991 , 0.8525598701134546
[[ 0.86503305]
 [-0.32593216]
 [-0.26029948]
 [-0.27880142]]

[[-0.38434776]
 [-0.59939273]
 [ 0.55944422]
 [ 0.42429627]]

[[0.29207197]
 [0.11682879]
 [0.36508996]
 [0.87621591]]

[[-0.12506269]
 [-0.0709455 ]
 [-0.59485825]
 [ 0.79086644]]


&#9989; <font color=red>**Question 3:**</font>  The vector associated with the eigenvalue of 1 is the steady state vector $S$. Create a new vector $\hat{S}$ that is equivlent to the steady state vector scaled by a constant $c$ (i.e. $cS = \hat{S}$) such that the components of this new vector sum up to one (i.e. $\sum_{i=0}^{n}\hat{S}_i = 1$). 

In [4]:
v = np.sum(v3)
c = 1/v
print(c)
S = c*v3
print(S)
print(np.sum(S))

0.6059847191423012
[[0.17699115]
 [0.07079646]
 [0.22123894]
 [0.53097345]]
1.0000000000000002


&#9989; <font color=red>**Question 4:**</font> Based on the long term steady state vector that you calculated, which page are people likely to spend the most time? Explain how you came up with the answer.

**The people are going to spend the most time in blog state because it has the largest value in the vector system.**

&#9989; <font color=red>**Question 5:**</font>  Construct a new matrix $P$ of the same size as $A$ with a copy of the normalized steady state vector ($\hat{S}$) in each column. 

In [5]:
# Put the answer to the above question here
P = np.matrix([[0.17699115,0.7079646,0.22123894,0.53097345],[0.17699115,0.7079646,0.22123894,0.53097345],[0.17699115,0.7079646,0.22123894,0.53097345],[0.17699115,0.7079646,0.22123894,0.53097345]]).T
sym.Matrix(P)

⎡0.17699115  0.17699115  0.17699115  0.17699115⎤
⎢                                              ⎥
⎢0.7079646   0.7079646   0.7079646   0.7079646 ⎥
⎢                                              ⎥
⎢0.22123894  0.22123894  0.22123894  0.22123894⎥
⎢                                              ⎥
⎣0.53097345  0.53097345  0.53097345  0.53097345⎦

&#9989; <font color=red>**Question 6:**</font> If exactly 10,000 people are on the website at a time (assume this is constant), estimate how many people will be on the About page over the long term?  Explain how you came up with your answer.

In [6]:
# Put the answer to the above question here
numberOfPeople = 10000*S[1]
print(numberOfPeople)

[[707.96460177]]


Therefore, number of people over the long term would be $708$.

----
Written by Dr. Dirk Colbry, Michigan State University
<a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>.

###STARTFOOTER###

-----
### Congratulations, we're done!


----